rasa_nlu/classifiers/embedding_intent_classifier.py

```diff
@@ -0,0 +1,565 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import logging
+import os
+
+import typing
+from future.utils import PY3
+from typing import List, Text, Any, Optional, Dict
+
+from rasa_nlu.classifiers import INTENT_RANKING_LENGTH
+from rasa_nlu.components import Component
+from rasa_nlu.config import RasaNLUModelConfig
+from rasa_nlu.model import Metadata
+from rasa_nlu.training_data import Message
+from rasa_nlu.training_data import TrainingData
+import numpy as np
+import tensorflow as tf
+
+try:
+    import cPickle as pickle
+except:
+    import pickle
+
+logger = logging.getLogger(__name__)
+
+if typing.TYPE_CHECKING:
+    import tensorflow as tf
+
+
+class EmbeddingIntentClassifier(Component):
+    """Intent classifier using supervised embeddings.
+
+    The embedding intent classifier embeds user inputs
+    and intent labels into the same space.
+    Supervised embeddings are trained by maximizing similarity between them.
+    It also provides rankings of the labels that did not "win".
+    The embedding intent classifier needs to be preceded by
+    a featurizer in the pipeline.
+    This featurizer creates the features used for the embeddings.
+
+    Based on the starspace idea from: https://arxiv.org/abs/1709.03856"""
+
+    name = "intent_classifier_tensorflow_embedding"
+
+    provides = ["intent", "intent_ranking"]
+
+    requires = ["text_features"]
+
+    defaults = {
+        # nn architecture
+        "num_hidden_layers_a": 2,
+        "hidden_layer_size_a": [256, 128],
+        "num_hidden_layers_b": 0,
+        "hidden_layer_size_b": [],
+        "batch_size": 32,
+        "epochs": 300,
+
+        # embedding parameters
+        "embed_dim": 10,
+        "mu_pos": 0.8,  # should be 0 < ... < 1 for 'cosine'
+        "mu_neg": -0.4,  # should be -1 < ... < 1 for 'cosine'
+        "similarity_type": 'cosine',  # should be 'cosine' or 'inner'
+        "num_neg": 10,
+        "use_max_sim_neg": True,  # flag which loss function to use
+
+        # regularization
+        "C2": 0.002,
+        "C_emb": 0.8,
+        "droprate": 0.2,
+
+        # flag if to tokenize intents
+        "intent_tokenization_flag": False,
+        "intent_split_symbol": '_'
+    }
+
+    def __init__(self, component_config=None,
```

Comment:
this method is suuuuper long. I'd put stuff like loading parameters from config and checking if hidden layer size is valid in separate functions and then call them here.